indykite/resource_consent.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2022 IndyKite
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indykite
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/indykite/indykite-sdk-go/config"
+	configpb "github.com/indykite/indykite-sdk-go/gen/indykite/config/v1beta1"
+)
+
+const (
+	consentPurposeKey        = "purpose"
+	consentApplicationIDKey  = "application_id"
+	consentValidityPeriodKey = "validity_period"
+	consentRevokeAfterUseKey = "revoke_after_use"
+	consentDataPointsKey     = "data_points"
+)
+
+func resourceConsent() *schema.Resource {
+	readContext := configReadContextFunc(resourceConsentFlatten)
+
+	return &schema.Resource{
+		CreateContext: configCreateContextFunc(resourceConsentBuild, readContext),
+		ReadContext:   readContext,
+		UpdateContext: configUpdateContextFunc(resourceConsentBuild, readContext),
+		DeleteContext: configDeleteContextFunc(),
+		Importer: &schema.ResourceImporter{
+			StateContext: basicStateImporter,
+		},
+
+		Timeouts: defaultTimeouts(),
+		Schema: map[string]*schema.Schema{
+			locationKey:    locationSchema(),
+			customerIDKey:  setComputed(customerIDSchema()),
+			appSpaceIDKey:  setComputed(appSpaceIDSchema()),
+			nameKey:        nameSchema(),
+			displayNameKey: displayNameSchema(),
+			descriptionKey: descriptionSchema(),
+			createTimeKey:  createTimeSchema(),
+			updateTimeKey:  updateTimeSchema(),
+
+			consentPurposeKey: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Purpose of the consent",
+			},
+			consentApplicationIDKey: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Application ID",
+			},
+			consentValidityPeriodKey: {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: "Specifies the duration in second that the consent remains valid, ranging from 1 day to 2 years",
+			},
+			consentRevokeAfterUseKey: {
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				Description: "If set to true, the consent will be revoked after it is used once",
+			},
+			consentDataPointsKey: {
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: "Data points is a list of properties related to the Digital twin that the consent is for",
+			},
```

Comment:
I'm not sure this is good solution.

1. The order in array matters. I'm not sure it is intended. But if you just change order in the list, Terraform will be unhappy
2. If you format JSON differently, Terraform will be unhappy again. You should implement `DiffSuppressFunc: structure.SuppressJsonDiff,` for element.